### ISSUE TITLE
missing wayland clipboard support

### ISSUE THREAD

**Issue by prahal** [input-leap/input-leap#1698]
### What happened?

Copying from an input-leap peer does not allow to paste to another one.

### Version

From Git HEAD or commit (specify below)

### Git commit hash (if applicable)

c6f097e9

### If applicable, where did you install Input Leap from?

Git

### What OSes are you seeing the problem on? (Check all that apply)

Linux

### If applicable, are you using Wayland or X11?

Any combination of Wayland and Xorg input-leap client and servers

### What OS versions are you using?

Debian trixie (both server and client)

### Relevant log output

```shell
When I copy nothing is logged on the barrier.log file on either server or client (while doing Ctrl+C). DEBUG2 level.
```

### Any other information

As told in https://github.com/input-leap/input-leap/issues/1673 by @shymega

> > Is the non-working clipboard sharing in wayland mode expected?

> Clipboard is a mess. It doesn't really work on X11 either. My position on it is a refactor.

Refactoring might have been discussed on IRC in April 2023 per https://github.com/input-leap/input-leap/discussions/1614, though no news if the discussion ever happened.

This is the biggest milestone left for Wayland support as far as I know https://github.com/input-leap/input-leap/issues/109
The libei support for Wayland inputs still has a few rough edges and is not supported on all Wayland compositors yet, but the framework is there.

**Comment by shymega** [comment]
Nothing was discussed on IRC, but another problem is the event handler. Currently, we use Qt's 'single slot' system. I discussed with @p12tic over email, and we came to a conclusion that the *Observer* pattern might work better, but that's basically single-slots, according to Wikipedia... so I'm not sure. I think we do need to refactor clipboard, and by extension, the event handler though.

However, we should make a first release first.

**Comment by whot** [comment]
I'm pretty sure we should be using the [Clipboard portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Clipboard) together with the RemoteDesktop/InputCapture sessions. Don't immediately know how the code will look like but that's the starting point.

**Comment by cobowers** [comment]
Just wanted to add a bump to this. Also seeing the same issue

**Comment by shymega** [comment]
> Just wanted to add a bump to this. Also seeing the same issue

Please, refrain from bumping issues! If you're experiencing the same issue, show you're affected by using the GitHub reactions feature on the first post. You also can subscribe to issues.

It doesn't help _anyone_, by adding a 'me too' comment, and creates noise.

**Comment by marcelopio** [comment]
Was looking on how this could be implemented.

According to xdg-desktop-portal documentation
> This portal does NOT create it’s own session. Instead, it offers existing sessions created from other portals the option to integrate with the clipboard. For whether this interface is supported for a given session, refer to that portal’s documentation. See [Remote Desktop](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.RemoteDesktop.html#org-freedesktop-portal-remotedesktop) to integrate clipboard with the remote desktop session.

So, from what I understand, our InputCapture session would need to request the clipboard before activating.

But the clipboard only works with RemoteDesktop sessions for now:
https://github.com/flatpak/xdg-desktop-portal/blob/3aba00899855d7afc517a4e2e8a22b41318b1e04/src/clipboard.c#L131

So, I think there should be a request to change that on the xdg-desktop-portal, and also a request to implement that on each platform implementation (GNOME, KDE, etc).

@whot is this correct?

**Comment by edgan** [comment]
Nikau seems to already have a working clipboard solution based on uinput for Wayland Gnome. I am using it right now. But it only supports keyboard shortcuts for switching, not edge flipping. It also isn't cross platform.

It seems like, as I have read elsewhere, the clipboard solutions for Wayland are going to depend on which compositor or desktop environment you are using. There is also the situation of where we are today versus where we may find ourselves in another few years. The input-leap project is already suffering from this "forced" migration from Xorg to Wayland, and all that work that requires. It seems like waiting for a complete or purer solution maybe stopping action here. I suggest going with a solution that can at least work for some people today instead of nothing. 

Based on the summaries below it seems like input-leap is one important feature away from being the clear choice. At this moment it seems like if you really want Wayland and Linux only Nikau is the clear source, and at some point in the future lan-mouse may take over.

input-leap:
GUI
Encryption
No clipboard for Wayland
Keyboard shortcuts and edge flipping(the better solution IMHO)
Linux, Windows, and MacOS support

nikau:
No GUI
Encryption
Clipboard for Wayland
Wayland Gnome, at least
Keyboard shortcut only
Linux only

lan-mouse:
GUI
No encryption, but has it on the roadmap
No clipboard, but has it on the roadmap
Wayland wlroots/KDE/Gnome support
Windows support(I think)
MacOS, input emulation but not input capture
Keyboard shortcut only(I think)

rkvm:
No GUI
Encryption
No clipboard. Currently claims not going to do it. Suggests using external programs to do it.
Keyboard shortcut only(I think)
Linux only

Clipboard syncers:
https://github.com/trueaniki/shared-clipboard
https://github.com/zachhuff386/clipsync
https://github.com/quackduck/uniclip

**Comment by whot** [comment]
@marcelopio - yeah, I think that's about correct


@edgan Nikau works in a fundamentally completely different and cannot directly be compared to input-leap/synergy. It uses uinput which makes it impossible to ever run in a sandbox, unlike input-leap where the whole point of adding portal and libei was to make that possible. And the [nikau website](https://nikau.nickbp.com/) claims:
```
Clipboard transfers currently rely on X11 clipboard APIs, which are supported by
all X11 WMs as well as some Wayland clients like GNOME. I plan to support
[wlroots clipboard APIs](https://wayland.app/protocols/wlr-data-control-unstable-v1)
 used by some other Wayland clients.
```

As with everything, the difficult bit is not necessarily the implementation but that someone needs to spend the time doing it.

**Comment by edgan** [comment]
@whot I get it works a completely different way, and that may have different issues. The point is end users don't care about the implementation, they just want something that works. Another way of what I was trying to say above, don't let perfect be the enemy of good. Having to wait years for a solution is just going to cause people to move on, one way or another.

I am finding myself even more between a rock and a hard place as a user. The one thing hard holding be back from Wayland is input-leap's lack of clipboard support, or some alternative. Yet I am now running into a 10 second delay issue with Xorg and Gnome on login. If I login with Wayland there is no such issue. So even just sticking with Xorg and input-leap has a different downside.

Ultimately, a lot like the recent Wayland protocol debate, we have various projects all trying to drive toward Wayland as the 100% replacement for Xorg, but the project as a whole just moves at a glacier pace. Input-leap's lack of clipboard support is a side effect. Part of my point is don't wait on them, workaround the problem.

**Comment by edgan** [comment]
I just ran across this issue for xdg-desktop-portal. The thread is nearly two years old, closed, and they basically say they aren't feeling a generic clipboard portal. They also say, why wouldn't something like input-leap just use the remotedesktop portal.

https://github.com/flatpak/xdg-desktop-portal/issues/743

**Comment by marcelopio** [comment]
I think they are more open now

https://github.com/flatpak/xdg-desktop-portal/discussions/1459

**Comment by whot** [comment]
> why wouldn't something like input-leap just use the remotedesktop portal.

IL *is* using the RD portal on the client side but to get clipboard access it requires the Clipboard portal, at least according to the documentation. IOW the bits for this exist, they just need to be hooked up. What is missing is the server-side because the InputCapture portal doesn't do clipboards but that can presumably be done in the same manner as the RD portal integration.

> Part of my point is don't wait on them, workaround the problem.

Working around a problem requires time. This time could be spent fixing the problem properly.

**Comment by darkblaze69** [comment]
> Yet I am now running into a 10 second delay issue with Xorg and Gnome on login

the PR is here https://github.com/input-leap/input-leap/pull/1084

it needs rebase

**Comment by whot** [comment]
> the PR is here https://github.com/input-leap/input-leap/pull/1084, it needs rebase

Original user is @ghost, i.e. deleted so your best chances are closing that PR, taking the patches, rebasing it locally and pushing a new PR superseding the existing one. That's more likely to ever get handled. (I'm not an IL maintainer but that's how I'd do it in any other project to speed things along)

**Comment by edgan** [comment]
> Working around a problem requires time. This time could be spent fixing the problem properly.

It is a question of time for each. If a workaround can be done in a month or two, and a proper solution takes a year or two, I think most people would start with the workaround. This is especially true when the solutions are likely to involve third party projects cooperation.

Speaking of workarounds I just explored some of the clipboard sharing software I linked above. I found clipsync allows bidirectionally copy and pasting between two Wayland systems. I don't fully understand how it works. It seems like it is Golang code wrapped around a little C code that calls some X libraries. My next thought was to make a PR to merge it's functionality into input-leap. Then I ran into #2001 which derails that being worth it.

**Comment by edgan** [comment]
> > Yet I am now running into a 10 second delay issue with Xorg and Gnome on login
> 
> the PR is here #1084
> 
> it needs rebase

I am not sure why you linked to that PR. It has nothing to do with my login problem. Also I have found a workaround for that issue today. I reset all Gnome settings and configuration by resetting my home directory. I have then just been pulling files back in from the backup.

**Comment by littleblack111** [comment]
> I just ran across this issue for xdg-desktop-portal. The thread is nearly two years old, closed, and they basically say they aren't feeling a generic clipboard portal. They also say, why wouldn't something like input-leap just use the remotedesktop portal.
> 
> [flatpak/xdg-desktop-portal#743](https://github.com/flatpak/xdg-desktop-portal/issues/743)

https://github.com/flatpak/xdg-desktop-portal/pull/852

**Comment by littleblack111** [comment]
no one interested doing this?

**Comment by marcelopio** [comment]
It is not implemented on xdg-desktop-portal.

https://github.com/input-leap/input-leap/issues/1698#issuecomment-2403549793

There is nothing that can be done on input-leap side until they implement it there, and than someone needs to do that on the compositor side, and than input-leap can implement

**Comment by shymega** [comment]
Exactly, nothing can be done here yet.

**Comment by prahal** [comment]
> > I just ran across this issue for xdg-desktop-portal. The thread is nearly two years old, closed, and they basically say they aren't feeling a generic clipboard portal. They also say, why wouldn't something like input-leap just use the remotedesktop portal.
> > [flatpak/xdg-desktop-portal#743](https://github.com/flatpak/xdg-desktop-portal/issues/743)
> 
> it seems to be added: [flatpak/xdg-desktop-portal#852](https://github.com/flatpak/xdg-desktop-portal/pull/852)
> 
> in org.freedesktop.portal.Clipboard

This is Remote Desktop portal support ie for the IL clients. No support for Input Capture portal (Input Leap server) as of yet.

**Comment by Amari-Mecheri** [comment]
I use Kde Connect on Fedora 41 KDE as a workaround.

**Comment by argoneus** [comment]
Added a PR to clarify this limitation in the README: https://github.com/input-leap/input-leap/pull/2210

**Comment by spiregarden** [comment]
Is this what we need?  https://github.com/deskflow/deskflow/issues/8031

If I understand correctly, the feature is implemented in [KDE Plasma 6.4](https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/merge_requests/337) and [GNOME 45](https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome/-/merge_requests/53)

**Comment by marcelopio** [comment]
Unfortunately, these implementations only tackles the Remote Desktop session type, not the Input Capture session type which is the one Input Leap and Deskflow uses, so this is a good progress, but doesn't progress this issue specifically at all, all implementations are still pending

- xdg-desktop-portal
- KDE/GNOME
- Input Leap
